network/p2p/scoring/registry_test.go

```diff
@@ -422,6 +423,71 @@ func TestPersistingInvalidSubscriptionPenalty(t *testing.T) {
 	assert.Equal(t, 0.0, record.Penalty) // penalty should be zero.
 }
 
+// TestSpamRecordDecayAdjustment ensures that spam record decay is increased each time a peers score reaches the scoring.IncreaseDecayThreshold eventually
+// sustained misbehavior will result in the spam record decay reaching the minimum decay speed .99, and the decay speed is reset to the max decay speed .8.
+func TestSpamRecordDecayAdjustment(t *testing.T) {
+	flowConfig, err := config.DefaultConfig()
+	require.NoError(t, err)
+	scoringRegistryConfig := flowConfig.NetworkConfig.GossipSubConfig.GossipSubScoringRegistryConfig
+	// increase configured DecayRateDecrement so that the decay time is increased faster
+	scoringRegistryConfig.DecayRateDecrement = .1
+	scoringRegistryConfig.DecayAdjustInterval = time.Second
+
+	peer1 := unittest.PeerIdFixture(t)
+	peer2 := unittest.PeerIdFixture(t)
+	reg, spamRecords := newScoringRegistry(
+		t,
+		scoringRegistryConfig,
+		withStakedIdentity(peer1),
+		withValidSubscriptions(peer1),
+		withStakedIdentity(peer2),
+		withValidSubscriptions(peer2))
+
+	// initially, the spamRecords should not have the peer ids.
+	assert.False(t, spamRecords.Has(peer1))
+	assert.False(t, spamRecords.Has(peer2))
+
+	// simulate sustained malicious activity from peer1, eventually the decay speed
+	// for a spam record should be reduced to the MinimumSpamPenaltyDecaySpeed
+	require.Eventually(t, func() bool {
+		reg.OnInvalidControlMessageNotification(&p2p.InvCtrlMsgNotif{
+			PeerID:  peer1,
+			MsgType: p2pmsg.CtrlMsgPrune,
+		})
+		record, err, ok := spamRecords.Get(peer1)
+		require.NoError(t, err)
+		require.True(t, ok)
+		return record.Decay == scoring.MinimumSpamPenaltyDecaySpeed
+	}, 5*time.Second, 500*time.Millisecond)
```

Comment:
This overall test is satisfactory, but a more detailed assessment would be beneficial. Specifically, a test that methodically verifies the gradual reduction of the decay speed from its maximum to minimum value, by checking the intermediate values, would enhance our testing granularity of the decay mechanism.

Reply:
https://github.com/onflow/flow-go/pull/4891/commits/eafd43476f3be027bcef5fa0c6041b61f44be8ee